src/mpi/init/initthread.c

```diff
@@ -76,120 +76,140 @@ int MPI_Init_thread(int *argc, char ***argv, int required, int *provided)
    how MPICH was configured. */
 extern const char MPII_Version_device[];
 
-int MPIR_Init_thread(int *argc, char ***argv, int required, int *provided)
+int MPIR_Init_thread(int *argc, char ***argv, int user_required, int *provided)
 {
     int mpi_errno = MPI_SUCCESS;
+    int required = user_required;
+    int err;
 
-    MPII_init_thread_and_enter_cs();
-    /* ---- Entered CS ------------------------------------------------ */
+    /**********************************************************************/
+    /* Section 1: base components that other components rely on.
+     * These need to be intialized first.  They have strong
+     * dependencies between each other, which makes it a pain to
+     * maintain them.  Hopefully, the number of such components is
+     * small. */
+    /**********************************************************************/
 
     MPL_wtime_init();
-    MPII_pre_init_dbg_logging(argc, argv);
-
-    mpi_errno = MPIR_T_env_init();
-    MPIR_ERR_CHECK(mpi_errno);
 
-    MPII_init_request();
+    MPID_Thread_init(&err);
```

Comment:
This one has the `MPID_` prefix, but it is not a MPID API. Isn't it a bit confusing? It will be so much easier to call it `MPII_Thread_init`. Device layer always can do their own thread-related init anyway.

This also links back to my petpeeve. We only use and need MPID wrapper for thread CS stuff, the `thread/mutex/cond` stuff are already wrapped in `MPL`, and device are simply using them. Having the extra MPID and MPIDU redirections seems pretty confusing.

Reply:
Right now, all thread stuff sits in the device, so it is a device-level function.  So `MPID_` is correct.  I would generally agree that there's not much special things that the device can do for threads, so it should probably not be `MPID_` (including thread CS stuff).  But that's a lot of code churn which will affect our downstream derivatives, which is why I avoided it.